src/lib.rs

```diff
@@ -141,15 +156,38 @@ impl SecretClient {
         A: for<'de> Deserialize<'de>,
     {
         let path = format!("{}/{}", self.base_path, path);
-        self.read_secrets_internal::<A>(&path).await
+        self.read_secrets_internal::<A>(&path, None).await
+    }
+
+    /// Read secrets with metadata.
+    #[cfg(feature = "read")]
+    pub async fn read_secrets_with_metadata<A>(
+        &self,
+        path: Option<&str>,
+    ) -> Result<SecretWithMetaData<A>>
+    where
+        A: for<'de> Deserialize<'de>,
+    {
+        let path = path.unwrap_or(&self.base_path);
+        let metadata: vaultrs::api::kv2::responses::ReadSecretMetadataResponse =
+            kv2::read_metadata(&self.client, &self.mount, path).await?;
+
+        let data = self
+            .read_secrets_internal(path, Some(metadata.current_version))
+            .await?;
+
+        Ok(SecretWithMetaData { data, metadata })
     }
 
     #[cfg(feature = "read")]
-    async fn read_secrets_internal<A>(&self, path: &str) -> Result<A>
+    async fn read_secrets_internal<A>(&self, path: &str, version: Option<u64>) -> Result<A>
```

Comment:
Ignore if you want but what do you think about a dedicated version enum? Works like this just had the impulse to add an enum somehow.

Reply:
The enum version that I can imagine would be bascially `Option` but with a fixed value type? Or do you have something else in mind?

Reply:
Yeah, does not make sense. As said, was just an impulse. it is internal anyways, so not worth it. 🚀